guarddog/scanners/pypi_project_scanner.py

```diff
@@ -97,49 +102,26 @@ def safe_parse_requirements(req):
                 valid_versions = None
                 project_exists_on_pypi = True
-                for spec in requirement.specs:
-                    qualifier, version = spec
-
-                    try:
-                        available_versions = versions(requirement.project_name)  # type: list[str]
-                    except Exception:
-                        sys.stderr.write(f"Package {requirement.project_name} not on PyPI\n")
-                        project_exists_on_pypi = False
-                        continue
-
-                    used_versions = None
-
-                    match qualifier:
-                        case ">":
-                            used_versions = {v for v in available_versions if v > version}
-                        case "<":
-                            used_versions = {v for v in available_versions if v < version}
-                        case ">=":
-                            used_versions = {v for v in available_versions if v >= version}
-                        case "<=":
-                            used_versions = {v for v in available_versions if v <= version}
-                        case "==":
-                            matches = [re.search(version, candidate) for candidate in available_versions]
-                            filtered_matches = list(filter(None, matches))
-                            str_matches = [v.string for v in filtered_matches]
-                            used_versions = set(str_matches)
-                        case "~=":
-                            prefix = "".join(version.split(".")[:-1])
-                            for available_version in available_versions:  # sorted decreasing
-                                if available_version >= version and available_version.startswith(prefix):
-                                    used_versions = set(available_version)
-                                    break
-                        case _:
-                            sys.stderr.write(f"Unknown qualifier: {qualifier}")
-                            continue
-
-                    if valid_versions is None:
-                        valid_versions = used_versions
-                    else:
-                        valid_versions = valid_versions & used_versions
+                try:
+                    available_versions = versions(
+                        requirement.project_name
+                    )  # type: list[str]
+                except Exception:
+                    sys.stderr.write(
+                        f"Package {requirement.project_name} not on PyPI\n"
+                    )
+                    project_exists_on_pypi = False
+                    continue
+
+                r = Requirement(str(requirement))
+
+                matched_versions = [m for m in r.specifier.filter(available_versions)]
+
+                if not VERIFY_ALL_DEPENDENCIES and matched_versions:
+                    matched_versions = [sorted(matched_versions).pop()]
```

Comment:
aren't you missing the last version due to the pop()? I think (not sure) that this should be:

```
matched_versions.extend(sorted(matched_versions))
```

Reply:
No, is the other way around, we're only keeping the last one.
i.e:
```
>>> x = {2,3,1}
>>> sorted(x)
[1, 2, 3]
>>> sorted(x).pop()
3
```

This is executed when `VERIFY_ALL_DEPENDENCIES` is **false**

Reply:
are we sure that `sorted` is the appropriate way to sort version strings?

Reply:
No:
```

>>> from semantic_version import Version # npm
>>> vmax = "1.10.0"
>>> vmin = "1.2.0"
>>> vmax > vmin
False
>>> Version(vmax) > Version(vmin)
True
...

>>> from packaging.version import Version # pypi
>>> Version(vmax) > Version(vmin)
True
```

I can address that by sorting the keys with the semver specs

Reply:
fixed in a99ed28495f1bcf111324fcb52e4cd0e074bb8c4